kit/score/score_test.go

```diff
@@ -32,3 +37,103 @@ func TestNormalize(t *testing.T) {
 		t.Errorf("Normalize(%d) = %d, expected %d", newMaxScore, sc.Score, expectedScore)
 	}
 }
+
+func TestScoreDetails(t *testing.T) {
+	tests := []struct {
+		name     string
+		messages []string
+		fatal    int // index of the message that should cause a fatal error
+		expected []string
+	}{
+		{
+			name:     "no messages",
+			messages: []string{},
+			fatal:    -1,
+			expected: []string{},
+		},
+		{
+			name:     "one message",
+			messages: []string{"first"},
+			fatal:    -1,
+			expected: []string{"first"},
+		},
+		{
+			name:     "two messages",
+			messages: []string{"first", "second"},
+			fatal:    1,
+			expected: []string{"first", "second"},
+		},
+		{
+			name:     "two messages",
+			messages: []string{"first", "second"},
+			fatal:    0,
+			expected: []string{"first"},
+		},
+		{
+			name:     "three messages",
+			messages: []string{"first", "second", "third"},
+			fatal:    1,
+			expected: []string{"first", "second"},
+		},
+	}
+
+	originalStdout := os.Stdout
+	r, w, _ := os.Pipe()
+	os.Stdout = w
+	defer w.Close()
+	wg := &sync.WaitGroup{}
+
+	// These tests simulate the behaviour in our test flow.
+	// i.e., code tested in Docker containers, where the output is captured and parsed.
+	// We expect that our helper (Errorf, Fatalf) methods will correctly add the test details to the score object.
+	// And that the parsing of the test details will be correct.
+	for _, test := range tests {
+		sc := &score.Score{}
+		wg.Add(1)
+		// run the test in a goroutine to avoid Fatalf
+		// from exiting the test via t.FailNow (runtime.Goexit)
+		subT := &testing.T{}
+		go func() {
+			defer wg.Done()
+			for i, m := range test.messages {
+				if test.fatal == i {
+					sc.Fatalf(subT, m)
+				} else {
+					sc.Errorf(subT, m)
+				}
+			}
+		}()
+		wg.Wait()
+		sc.Print(t)
+		out := make([]byte, 1024)
+		n, _ := r.Read(out)
+
+		parsedScore := &score.Score{}
+		if err := json.Unmarshal(out[:n], parsedScore); err != nil {
+			t.Fatalf("Failed to unmarshal score: %v", err)
+		}
+
+		gotMessages := parseTestDetails(parsedScore.TestDetails)
```

Comment:
Can't we also check the `dir/file.go:123: ` part? One approach could be to filter the start of the path and only check the last part of it: `<filtered path>/quickfeed/kit/score/score_test.go:109`. It will require that we fix the line numbers if we update the test. It can be a hassle if we change it a lot... However, we at least check that the line numbers get extracted correctly by the `parseTestDetails` function, without having to write a separate test for it. If we do used fix line numbers, then we should have a comment saying that they must be updated if the line numbers change.